pandas/core/arrays/floating.py

```diff
@@ -0,0 +1,641 @@
+    if mask.any():
+        values = values.copy()
+        values[mask] = np.nan
+        values = values.astype(dtype, copy=False)  # , casting="safe")
+    else:
+        values = values.astype(dtype, copy=False)  # , casting="safe")
+
+    return values, mask
+
+
+class FloatingArray(BaseMaskedArray):
+    """
+    Array of floating (optional missing) values.
+
+    .. versionadded:: 1.1.0
+
+    .. warning::
+
+       FloatingArray is currently experimental, and its API or internal
+       implementation may change without warning. Expecially the behaviour
+       regarding NaN (distinct from NA missing values) is subject to change.
+
+    We represent a FloatingArray with 2 numpy arrays:
+
+    - data: contains a numpy float array of the appropriate dtype
+    - mask: a boolean array holding a mask on the data, True is missing
+
+    To construct an FloatingArray from generic array-like input, use
+    :func:`pandas.array` with one of the float dtypes (see examples).
+
+    See :ref:`integer_na` for more.
+
+    Parameters
+    ----------
+    values : numpy.ndarray
+        A 1-d float-dtype array.
+    mask : numpy.ndarray
+        A 1-d boolean-dtype array indicating missing values.
+    copy : bool, default False
+        Whether to copy the `values` and `mask`.
+
+    Attributes
+    ----------
+    None
+
+    Methods
+    -------
+    None
+
+    Returns
+    -------
+    FloatingArray
+
+    Examples
+    --------
+    Create an FloatingArray with :func:`pandas.array`:
+
+    >>> pd.array([0.1, None, 0.3], dtype=pd.Float32Dtype())
+    <FloatingArray>
+    [0.1, <NA>, 0.3]
+    Length: 3, dtype: Float32
+
+    String aliases for the dtypes are also available. They are capitalized.
+
+    >>> pd.array([0.1, None, 0.3], dtype="Float32")
+    <FloatingArray>
+    [0.1, <NA>, 0.3]
+    Length: 3, dtype: Float32
+    """
+
+    # The value used to fill '_data' to avoid upcasting
+    _internal_fill_value = 0.0
+
+    @cache_readonly
+    def dtype(self) -> _FloatingDtype:
+        return _dtypes[str(self._data.dtype)]
+
+    def __init__(self, values: np.ndarray, mask: np.ndarray, copy: bool = False):
+        if not (isinstance(values, np.ndarray) and values.dtype.kind == "f"):
+            raise TypeError(
+                "values should be floating numpy array. Use "
+                "the 'pd.array' function instead"
+            )
+        super().__init__(values, mask, copy=copy)
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy: bool = False) -> "FloatingArray":
+        values, mask = coerce_to_array(scalars, dtype=dtype, copy=copy)
+        return FloatingArray(values, mask)
+
+    @classmethod
+    def _from_sequence_of_strings(
+        cls, strings, dtype=None, copy: bool = False
+    ) -> "FloatingArray":
+        scalars = to_numeric(strings, errors="raise")
+        return cls._from_sequence(scalars, dtype, copy)
+
+    _HANDLED_TYPES = (np.ndarray, numbers.Number)
```

Comment:
Need to think about ops involving a mix of our extension arrays something like

```python
np.add(FloatingArray, IntegerArray)
```

How feasible is an `__array_ufunc__` on a base class of FloatingArray & IntegerArray?